Adding API For QnAMaker v4.0

bingisbestest opened the pull request:
Adding API spec for QnAMaker v4.0

Commit:
adding v4

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=4377) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=4377) before we can accept your contribution.<br/><br/>:x: bingisbestest [sign now](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=4377)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-rest-api-specs?pullRequest=4377) it.</sub>

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-java

azuresdkci commented:
Can one of the admins verify this patch?

yangyuan commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker-Api.json:
```diff
@@ -0,0 +1,1041 @@
+  "x-generator": "NSwag v11.13.2.0 (NJsonSchema v9.10.21.0 (Newtonsoft.Json v11.0.0.0))",
+  "swagger": "2.0",
+  "info": {
+    "title": "QnA Maker REST API",
```
I think should be "QnAMaker Client"

The title will be as client class name

yangyuan commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker-Api.json:
```diff
@@ -0,0 +1,1041 @@
+    "description": "QnA Maker REST API V4.0",
+    "version": "4.0"
+  },
+  "host": "westus.api.cognitive.microsoft.com",
```
Please use endpoint template (please use Face as reference)

yangyuan commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker-Api.json:
```diff
@@ -0,0 +1,1041 @@
+{
```
Please rename to QnAMaker.json

yangyuan commented on specification/cognitiveservices/data-plane/QnAMaker/readme.md:
````diff
@@ -0,0 +1,41 @@
+Configuration for generating QnAMaker Api SDK.
+
+``` yaml
+tag: QnAMaker
````
tag is generally something like "release_4_0"

yangyuan commented on specification/cognitiveservices/data-plane/QnAMaker/readme.md:
````diff
@@ -0,0 +1,41 @@
+These settings apply only when `--csharp` is specified on the command line.
+
+``` yaml $(csharp)
+csharp:
````
should indicate the location of swagger file. otherwise the autorest won't work

bingisbestest replied:
used face readme, let me know if its fine now

yangyuan reviewed (changes requested):
I left a few comments, because the autorest is not picking up the swagger so there might be more issues there.

Commit:
Addressing comments

Commit:
Use Endpoint Template

Commit:
Switch to parameter

Commit:
Formatting

Commit:
linter method issue

Commit:
examples ALTS

Commit:
example keys

Commit:
ops example

Commit:
Add kb example

Commit:
Adding Create Kb Example

Commit:
Download KB Example

yangyuan commented on specification/cognitiveservices/data-plane/QnAMaker/readme.md:
```diff
@@ -0,0 +1,99 @@
+  sync-methods: None
+  license-header: MICROSOFT_MIT_NO_VERSION
+  azure-arm: false
+  namespace: Microsoft.Azure.CognitiveServices.QnAMaker
```
namespace should be current `Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker`

yangyuan commented on specification/cognitiveservices/data-plane/QnAMaker/readme.md:
```diff
@@ -0,0 +1,99 @@
+  license-header: MICROSOFT_MIT_NO_VERSION
+  azure-arm: false
+  namespace: Microsoft.Azure.CognitiveServices.QnAMaker
+  output-folder: $(csharp-sdks-folder)/CognitiveServices/dataPlane/QnAMaker/Generated
```
$(csharp-sdks-folder)/CognitiveServices/dataPlane/Knowledge/QnAMaker/Generated

yangyuan reviewed (changes requested):
Found an issue in namespace and folder. Others look good

Commit:
Fix Namespaces

bingisbestest commented:
Gentle reminder

yangyuan reviewed (approved):
LGTM, please ping me in the C# SDK PR.

bingisbestest commented:
Sure. QQ: do i need to join a team to merge this request ? It says "You’re not authorized to merge this pull request.". Let me know. Thanks.

yangyuan commented:
@bingisbestest, ADX team can merge this.
@marstr was assigned to PR but he added "Reassign", IDK why.

sarangan12 commented:
Synched up offline with @veronicagg. Since I will be OOF starting tomorrow, @veronicagg will be taking over this PR. Assigned it to her.

Commit:
model validation errors

Commit:
change to octet stream

Commit:
add consume/produces

Commit:
remove response body

Commit:
fix set alt example

bingisbestest commented:
Fixed model validator errors.

salameer commented:
@veronicagg can we kindly expedite the review here?

veronicagg commented on specification/cognitiveservices/data-plane/QnAMaker/readme.md:
```diff
@@ -0,0 +1,99 @@
+  - repo: azure-sdk-for-python
+  - repo: azure-sdk-for-java
+  - repo: azure-sdk-for-go
+  - repo: azure-sdk-for-js
```
looks like js and node SDKs configuration is missing, if you look at the PR comments, there's no generated SDK PR out of those, please take a look at https://github.com/Azure/azure-rest-api-specs/tree/master/specification/sql/resource-manager to add readme.md files for js and node to configure SDK generation for them.

veronicagg replied:
I would recommend to take a look at a PR that got generated for one of the language SDKs (see PR comments), to check that the generated code looks as you'd like it to be.

bingisbestest replied:
done

bingisbestest replied:
ruby sdk fails. can you please take a look ?

veronicagg commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker.json:
```diff
@@ -0,0 +1,1275 @@
+          "Alterations"
+        ],
+        "summary": "Download alterations from runtime.",
+        "operationId": "Download Alterations",
```
check/update operationId

veronicagg commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker.json:
```diff
@@ -0,0 +1,1275 @@
+          "200": {
+            "description": "Collection of knowlegebases.",
+            "schema": {
+              "$ref": "#/definitions/KnowledgebasesDTO"
```
for these "get all"/ "list" operations, have you looked at x-ms-pageable? https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-pageable 
I'm asking in case you don't know about it and your model/service want to support it that way, doesn't mean you need to change it here.

bingisbestest replied:
Thanks we dont support paging anywhere

veronicagg commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker.json:
```diff
@@ -0,0 +1,1275 @@
+            },
+            "x-nullable": true,
+            "headers": {
+              "Location": {
```
since this is data plane, not sure if you're doing a long running operation this similar to ARM, if so, you can check out https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-long-running-operation

bingisbestest replied:
good idea I looked at it but our api is returning relative path... Is this required ? we would need to make changes to our api to return absolute path.

veronicagg replied:
don't worry then, I just thought I would share in case it was helpful

veronicagg commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker.json:
```diff
@@ -0,0 +1,1275 @@
+    "KnowledgebaseEnvironment": {
+      "type": "string",
+      "description": "Enumeration of knowledgebase environments.",
+      "x-enumNames": [
```
please take a look at https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-enum and use "modelAsString" : true if you think this list may expand later (to reduce breaking changes in generated SDKs)

bingisbestest replied:
good suggestion

veronicagg commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker.json:
```diff
@@ -0,0 +1,1275 @@
+    "ErrorCode": {
+      "type": "string",
+      "description": "Human readable error code.",
+      "x-enumNames": [
```
See if https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-enum would be suitable in this case

veronicagg commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker.json:
```diff
@@ -0,0 +1,1275 @@
+    "OperationState": {
+      "type": "string",
+      "description": "Enumeration of operation states.",
+      "x-enumNames": [
```
https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-enum

bingisbestest commented on specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker.json:
```diff
@@ -0,0 +1,1275 @@
+          "EndpointKeys"
+        ],
+        "summary": "Re-generates an endpoint key.",
+        "operationId": "Refresh Endpoint Keys",
```
Using Endpoint_RefreshKeys

Commit:
Update Operation Ids

Commit:
Fix case of description

Commit:
Enums as strings

Commit:
Adding md files for other langs

Commit:
Update Ruby readme

veronicagg commented:
@bingisbestest thanks for addressing the comments. 
Question: is there a reason why you left x-enumsNames in the spec?
Rest looks good. 
Ruby SDK is failing, but looks to be a known issue on our side.
Has this service been deployed on your side? If so, let me know and I can merge the PR.

Commit:
remove x-enumNames

bingisbestest commented:
Yes this service is already live

bingisbestest commented:
> @bingisbestest thanks for addressing the comments.
> Question: is there a reason why you left x-enumsNames in the spec?
> Rest looks good.
> Ruby SDK is failing, but looks to be a known issue on our side.
> Has this service been deployed on your side? If so, let me know and I can merge the PR.

Removed. Yes this service is live for some time now.
